File: ruby/test/wire/conversations_test.rb
# frozen_string_literal: true

require_relative "wiremock_test_case"

class ConversationsWireTest < WireMockTestCase
  def setup
    super

    @client = Sikaru::Client.new(
      api_key: "<token>",
      base_url: WIREMOCK_BASE_URL
    )
  end

  def test_conversations_list_messages_with_wiremock
    test_id = "conversations.list_messages.0"

    @client.conversations.list_messages(
      project_id: "project_id",
      conversation_id: "conversation_id",
      account_id: "account_id",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "conversations.list_messages.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "GET",
      url_path: "/v1/projects/project_id/conversations/conversation_id/messages",
      query_params: { "account_id" => "account_id" },
      expected: 1
    )
  end

  def test_conversations_record_message_with_wiremock
    test_id = "conversations.record_message.0"

    @client.conversations.record_message(
      project_id: "project_id",
      conversation_id: "conversation_id",
      account_id: "account_id",
      content: "content",
      delivered_at: "2024-01-15T09:30:00Z",
      message_id: "messageId",
      position: 1,
      role: "user",
      request_options: {
        additional_headers: {
          "X-Test-Id" => "conversations.record_message.0"
        }
      }
    )

    verify_request_count(
      test_id: test_id,
      method: "POST",
      url_path: "/v1/projects/project_id/conversations/conversation_id/messages",
      query_params: { "account_id" => "account_id" },
      expected: 1
    )
  end
end

File: rust/tests/conversations_test.rs
use sikaru::prelude::*;

mod wire_test_utils;

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_conversations_list_messages_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = Sikaru::new(config).expect("Failed to build client");

    let result = client
        .conversations
        .list_messages(
            &"project_id".to_string(),
            &"conversation_id".to_string(),
            &ListMessagesQueryRequest {
                account_id: "account_id".to_string(),
                limit: None,
                cursor: None,
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count(
        "GET",
        "/v1/projects/project_id/conversations/conversation_id/messages",
        Some(HashMap::from([(
            "account_id".to_string(),
            json!("account_id"),
        )])),
        1,
    )
    .await
    .unwrap();
}

#[tokio::test]
#[allow(unused_variables, unreachable_code)]
async fn test_conversations_record_message_with_wiremock() {
    wire_test_utils::reset_wiremock_requests().await.unwrap();
    let wiremock_base_url = wire_test_utils::get_wiremock_base_url();

    let mut config = ClientConfig {
        token: Some("<token>".to_string()),
        ..Default::default()
    };
    config.base_url = wiremock_base_url.to_string();
    let client = Sikaru::new(config).expect("Failed to build client");

    let result = client
        .conversations
        .record_message(
            &"project_id".to_string(),
            &"conversation_id".to_string(),
            &DeliveredMessage {
                account_id: "account_id".to_string(),
                content: "content".to_string(),
                delivered_at: DateTime::parse_from_rfc3339("2024-01-15T09:30:00Z").unwrap(),
                message_id: "messageId".to_string(),
                position: 1,
                role: DeliveredMessageRole::User,
                environment: None,
                trace_id: None,
            },
            None,
        )
        .await;

    assert!(result.is_ok(), "Client method call should succeed");

    wire_test_utils::verify_request_count(
        "POST",
        "/v1/projects/project_id/conversations/conversation_id/messages",
        Some(HashMap::from([(
            "account_id".to_string(),
            json!("account_id"),
        )])),
        1,
    )
    .await
    .unwrap();
}
